### DISCUSSION TITLE
Add XPath support

### DISCUSSION THREAD

**Discussion by drmohundro** [drmohundro/SWXMLHash#9] [Ideas]
Research adding XPath support. I suspect that the code could wrap `NSXMLDocument` to provide this functionality.

**Comment by tkrajacic** [comment]
Great idea to do it that way. Shouldn't be too hard then. Well, the devil's in the details probably ;)

**Comment by drmohundro** [comment]
My initial thought on an API is something like:

``` swift
let xml = SWXMLParse.xpath(xmlContent, xpathQuery)
// with a definition of xpath(xmlContent: String, xpathQuery: String) -> XMLIndexer
```

A benefit would be avoiding loading all of the XML into memory, but only what was initially returned from the XPath query. Then typical parsing could occur with what was returned via indexing or whatever.

Even better, though, it might make sense to add an xpath method on `XMLIndexer` itself. I suspect it'd need to be an XPath query that was relative to that element. I'll do some more digging on this.

**Comment by tkrajacic** [comment]
Ideally, the entire enum would be lazily constructed in any case.
So only accessing/requesting a certain element/node would parse the xml data.
Not sure if this is possible though.
This should probably another issue on its own?

**Comment by drmohundro** [comment]
At the moment, it parses the entire XML document into memory when `parse` is called. I almost put a disclaimer in the README that said to consider falling back to `NSXMLParser` if a large XML file were being parsed to avoid memory pressure!

It would definitely be better to do the parsing lazily, though. I don't think it would have to actually start parsing until the `element` property were accessed.

**Comment by cezheng** [comment]
`NSXMLDocument` is only available on OSX. For watchOS & iOS I guess the most easy way to get XPath is to wrap libxml2. And that's what most XML libraries on iOS do.

**Comment by drmohundro** [comment]
Great point - I hadn't dug in far enough to see that... I think you're right that libxml2 would probably be the better move then.

**Comment by onmyway133** [comment]
Sorry for hijacking into this, for XPath support, it's good to based on libxml2, which has excellent implementation of XPath, something like https://github.com/onmyway133/Reindeers
